### ISSUE TITLE
Support different data-models in Single-/MultiSelect components

### ISSUE THREAD

**Issue by yusijs** [equinor/design-system#1416]
In 11/10 cases our selectionlists are based on name+id models `{ "name": "Select me", "id": "long-ass-guid" }`, but EDS only has support for simple strings. 

**Describe the solution you'd like**
Being able to send in a different datatype, or to send in an Option render-prop where we can connect the id to the name ourselves.

**Comment by mimarz** [comment]
Thanks for suggesting this feature! 

We have planned a few improvements and tweaks for Select-components in the not to far future, so maybe we can get this done then as-well :)

**Comment by mimarz** [comment]
I have started working on this feature @yusijs 

I am thinking maybe a predefined data model (name, id) with and optionLabel render? This will make it easier to provide types and easier to maintain i think

We are also planning on making some changes to our "Select" lineup, where we are introducing [Combobox](https://eds-storybook-react-development.azurewebsites.net/?path=/docs/components-combobox--default) as the culmination of `SingleSelect` & `MultiSelect`. 

Multiple selects will be then toggleable with a `multiple` prop and data provided as an array of objects via a prop. This will let us also introduce new features aswell as #1430.

If all goes as planned `SingleSelect` & `MultiSelect` will eventually be phased out and replaced by `Combobox` 😅  Reason for this being is that the two component shared a lot of the same logic and keeping the develop them in parallel while still being identical would be to much unneeded maintenance overhead.

**Comment by yusijs** [comment]
First off, I love the simplification of one component with a prop controlling the checkboxes! 

Re. the data-model - I would personally prefer being able to either 

a) control the list of options myself, something like:
```
<Combobox>
  {opts.map(option => <Option value={option.id}>{option.name}</Option>)}
</Combobox>
```

or b) pass in a prop
`<Combobox optionLabel={opt => <span>opt.name</opt>} optionValue="id">`


option A is my personal wish, but I get that that is both extra work and maybe not the desired solution for everyone. I think having a predefined model with name, id will become to limiting in the long run (e.g if you depend on an external API, you would have to create your own datastructure on top)

**Comment by mimarz** [comment]
> First off, I love the simplification of one component with a prop controlling the checkboxes!
> 
> Re. the data-model - I would personally prefer being able to either
> 
> a) control the list of options myself, something like:
> 
> ```
> <Combobox>
>   {opts.map(option => <Option value={option.id}>{option.name}</Option>)}
> </Combobox>
> ```
> 
> or b) pass in a prop
> `<Combobox optionLabel={opt => <span>opt.name</opt>} optionValue="id">`
> 
> option A is my personal wish, but I get that that is both extra work and maybe not the desired solution for everyone. I think having a predefined model with name, id will become to limiting in the long run (e.g if you depend on an external API, you would have to create your own datastructure on top)

**Option A** could work, but that would mean we would have to parse children to get the data, and that could also be limiting in a way? What value/element do you get in the when an option is selected? We are also kinda limited to what downshift can do here.  I do also prefer this approach, its definitely cleaner and lets the developers chose their own datamodell, which is like.

**Option B** is deffo the safest and easiest approach. the predefined model doesn't need to be limiting, apart from required props, I think you can add whichever properties you want if the type is (correctly typed). True that its limiting in that you need to translate the apps datamodell to combobox datamodell tho...

I'll do some testing on Option A and see how that works. The `Option` could have a required `value`/`id` with optional `disabled` then. Question is how downshift would handle the "selection" part here 🤔

**Comment by yusijs** [comment]
On mobile, but wouldnt something like [this example ](https://codesandbox.io/embed/github/kentcdodds/downshift-examples/tree/master/?module=%2Fsrc%2Fother-examples%2Fdropdown-select-by-key%2FCustomDropdown%2Findex.js&codemirror=1)  work?

**Comment by mimarz** [comment]
> On mobile, but wouldnt something like [this example ](https://codesandbox.io/embed/github/kentcdodds/downshift-examples/tree/master/?module=%2Fsrc%2Fother-examples%2Fdropdown-select-by-key%2FCustomDropdown%2Findex.js&codemirror=1) work?

Yeah, I've used a lot of those examples for inspiration :) I don't think your link worked. Which example did you think of? @yusijs

**Comment by yusijs** [comment]
Meh, codesandbox was ass on mobile. :P I was thinking of `/src/downshift/other-examples/dropdown-select-by-key` [github](https://github.com/downshift-js/downshift-examples/tree/master/src/downshift/other-examples/dropdown-select-by-key)

It seems like a hybrid of options A and B. I struggled to get it running in codesandbox, but based on reading through it it seems that way at least.. :)

**Comment by mimarz** [comment]
> Meh, codesandbox was ass on mobile. :P I was thinking of `/src/downshift/other-examples/dropdown-select-by-key` [github](https://github.com/downshift-js/downshift-examples/tree/master/src/downshift/other-examples/dropdown-select-by-key)
> 
> It seems like a hybrid of options A and B. I struggled to get it running in codesandbox, but based on reading through it it seems that way at least.. :)

We are using the `downshift` hooks, so its kinda same, kinda not. 

Either way, I got around to testing Option A and it gets cunky and messy, as we need to have controll over what options are rendered. Sure we could extract the "data" from the `Option` elements sent in as children, but that just feels superfluous and unecessary when you can just send in a array of objects conveying the same information.

**Comment by mimarz** [comment]
Solved in #1600
